### PR TITLE
Hrbe 51

### PR DESCRIPTION
**Pull request by Vasantgoudapatil**
Screenshot 2023-10-12 at 6 47 26 PM

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[[source]]
+url = "https://pypi.org/simple"
+verify_ssl = true
+name = "pypi"
+
+[packages]
+django = "*"
+
+[dev-packages]
```
Comment (Abishek05) [inline Pipfile:9]: Shouldn't this file be not pushed? Please add to it.gitignore

File: accounts/user_manager.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        user=self.model(username=username,email=self.normalize_email(email))
+        user.set_password(password)
+        user.save()
+        return user
+        
+    def create_superuser(self, username,email,password=None):
+        
+        if password is None:
+            raise ValueError("Password should not be none")
+
```
Comment (Abishek05) [inline accounts/user_manager.py:21]: This file is not required

File: accounts/serializers.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+class AgencySerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Agency
+        fields = [
+            'name',
+            'website',
+            'phone',
+            'email',
+        ]
```
Comment (Abishek05) [inline accounts/serializers.py:12]: Agency was not asked as per the ticket

File: accounts/models.py [hunk 4]
```diff
@@ -58,9 +57,9 @@ def __str__(self):
         return str(self.name)
 
 
-class User(AbstractUser):
+class User(AbstractBaseUser,PermissionsMixin):
+    id = models.UUIDField(primary_key=True,default=uuid.uuid4, editable=False, unique=True)
```
Comment (Abishek05) [inline accounts/models.py:61]: Rever back to AbstractUser

File: accounts/models.py [hunk 5]
```diff
@@ -43,8 +43,7 @@
 
 
 class Agency(models.Model):
-    pk_id = models.BigAutoField(primary_key=True, editable=False)
-    id = models.UUIDField(default=uuid.uuid4, editable=False, unique=True)
+    id = models.UUIDField(primary_key=True,default=uuid.uuid4, editable=False, unique=True)
```
Comment (Abishek05) [inline accounts/models.py:46]: Same here, agency is out of scope of this ticket

File: accounts/models.py [hunk 6]
```diff
@@ -3,8 +3,8 @@
 from _decimal import Decimal
 from django.core.validators import RegexValidator, MinValueValidator
 from django.db import models
-from django.contrib.auth.models import AbstractUser
-
+from django.contrib.auth.models import AbstractUser,AbstractBaseUser,PermissionsMixin
+from .user_manager import UserManager
```
Comment (Abishek05) [inline accounts/models.py:7]: Writing own UserManager is unnecessary

File: accounts/migrations/0001_initial.py [hunk 7]
```diff
@@ -1,68 +1,179 @@
...
+                    models.CharField(
+                        blank=True,
+                        max_length=15,
+                        null=True,
+                        validators=[
+                            django.core.validators.RegexValidator(
+                                message="Enter Phone Number in the format +91. Upto 15 digits allowed",
+                                regex="^\\+?1?\\d{10,14}$",
+                            )
+                        ],
```
Comment (Abishek05) [inline accounts/migrations/0001_initial.py:43]: Do not edit the migrations file as it is not meant to be edited

File: accounts/views.py [hunk 8]
```diff
@@ -1,3 +1,20 @@
...
+    
+    def post(self,request):
+        user=request.data
+        serializer=self.serializer_class(data=user)
+        serializer.is_valid(raise_exception=True)
+        User.objects.create_user(username=user['username'],email=user['email'], password=user['password'])
+        
+        user_data=serializer.data
+        
+        return Response(user_data,status=status.HTTP_201_CREATED)
```
Comment (Abishek05) [inline accounts/views.py:20]: Use ModelViewset and not GenericAPIView

File: mysite/urls.py [hunk 9]
```diff
@@ -15,8 +15,9 @@
     2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
 """
 from django.contrib import admin
-from django.urls import path
+from django.urls import path,include
 
 urlpatterns = [
     path('admin/', admin.site.urls),
+    path('api/', include('accounts.urls')),
```
Comment (Abishek05) [inline mysite/urls.py:22]: Add routers

**Review by Abishek05 — CHANGES_REQUESTED** [review]
Changes to be made, many things are worked on are out of scope of the ticket

File: accounts/models.py [hunk 10]
```diff
@@ -5,52 +5,55 @@
...
+    ("TIMESHEET-PAYMENT", "Timesheet-payment"),
 )
 
 
 class Agency(models.Model):
-    pk_id = models.BigAutoField(primary_key=True, editable=False)
-    id = models.UUIDField(default=uuid.uuid4, editable=False, unique=True)
+    id = models.UUIDField(
+        primary_key=True, default=uuid.uuid4, editable=False, unique=True
+    )
```
Comment (Abishek05) [inline accounts/models.py:47]: As discussed, this line of code needs to be reverted

File: accounts/models.py [hunk 11]
```diff
@@ -60,21 +63,42 @@ def __str__(self):
 
 
 class User(AbstractUser):
+    id = models.UUIDField(
+        primary_key=True, default=uuid.uuid4, editable=False, unique=True
+    )
     first_name = models.CharField(max_length=150)
-    middle_name = models.CharField(max_length=150)
```
Comment (Abishek05) [inline accounts/models.py:64]: middle_name is required

File: accounts/models.py [hunk 12]
```diff
@@ -60,21 +63,42 @@ def __str__(self):
...
+    role = models.CharField(choices=USER_ROLE_CHOICES, max_length=30, default="CLIENT")
+    type = models.CharField(
+        choices=USER_TYPE_CHOICES, max_length=30, default="INDIVIDUAL"
+    )
+    agency = models.OneToOneField(
+        Agency, on_delete=models.CASCADE, blank=True, null=True
+    )
     is_email_verified = models.BooleanField(default=False)
     is_root_user = models.BooleanField(default=False)
+    is_profile_complete = models.BooleanField(default=False)
```
Comment (Abishek05) [inline accounts/models.py:93]: This is out of scope of this ticket, please remove `is_profile_complete` field

File: accounts/models.py [hunk 13]
```diff
@@ -60,21 +63,42 @@ def __str__(self):
...
+
+    USERNAME_FIELD = "email"
+    REQUIRED_FIELDS = ["username", "password"]
+
+    def save(self, *args, **kwargs):
+        if not self.pk:
+            self.is_root_user = True
+        super().save(*args, **kwargs)
```
Comment (Abishek05) [inline accounts/models.py:101]: Instead of this, make `is_root_user` default to true

File: accounts/utils.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+        try:
+            email = EmailMessage(
+                subject=data["email_subject"],
+                body=data["email_body"],
+                to=[data["to_email"]],
+            )
+            email.send()
+            print("Email sent successfully")
+        except Exception as e:
+            print(f"Email sending failed: {str(e)}")
```
Comment (Abishek05) [inline accounts/utils.py:16]: Remove debug messages

File: accounts/views.py [hunk 15]
```diff
@@ -1,8 +1,78 @@
...
+            user.is_email_verified = True
+            user.save()
+            return redirect("login")
+        except Exception as e:
+            return Response(
+                {"error": "Token verification failed"},
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+
```
Comment (Abishek05) [inline accounts/views.py:75]: Do not use APIView as discussed earlier. This is out of Scope. Please remove or comment out for now

File: accounts/views.py [hunk 16]
```diff
@@ -1,8 +1,78 @@
...
 from .models import Skill
 from .serializers import SkillSerializer
 from .permissions import IsAdminOrReadOnlyCustom
 
+
+class RegisterViewSet(viewsets.ModelViewSet):
+    queryset = User.objects.all()
+    serializer_class = RegisterSerializer
+
+    def post(self, request):
```
Comment (Abishek05) [inline accounts/views.py:25]: post() will not work in ModelViewset, you need to overwrite create() method

File: accounts/views.py [hunk 17]
```diff
@@ -1,8 +1,78 @@
...
+        serializer = self.serializer_class(data=user)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+        user_data = serializer.data
+        user = User.objects.get(email=user_data["email"])
+        token = RefreshToken.for_user(user).access_token
+        current_site = get_current_site(request)
+        relative_link = reverse("verify-email")
+
+        absurl = "http://" + current_site + relative_link + "?token=" + str(token)
```
Comment (Abishek05) [inline accounts/views.py:36]: use f strings here to presentation looks easier on the eyes

File: accounts/admin.py [hunk 18]
```diff
@@ -1,6 +1,7 @@
...
 from django.contrib import admin
-from .models import Skill
+from django.contrib.auth.admin import UserAdmin
+from .models import User, Skill
 
-# Register your models here.
-
-admin.site.register(Skill)
\ No newline at end of file
+admin.site.register(User, UserAdmin)
```
[the remark this answers is not in the document]
Comment (Vasantgoudapatil) [reply accounts/admin.py:6]: Ok will take that off

File: accounts/models.py (hunk 11 above)
Comment (Vasantgoudapatil) [reply accounts/models.py:64]: ok

File: accounts/models.py (hunk 13 above)
Comment (Vasantgoudapatil) [reply accounts/models.py:101]: ok

File: accounts/models.py (hunk 12 above)
Comment (Vasantgoudapatil) [reply accounts/models.py:93]: ok

File: accounts/utils.py (hunk 14 above)
Comment (Vasantgoudapatil) [reply accounts/utils.py:16]: ok

File: accounts/views.py (hunk 15 above)
Comment (Vasantgoudapatil) [reply accounts/views.py:75]: Ok will comment it out for now as it is required in another ticket

**Review by Abishek05 — APPROVED** [review]
LGTM

**Pull request merged**